tests/flytekit/unit/extras/tasks/test_shell.py

```diff
@@ -100,70 +100,104 @@ def test_input_output_substitution_files():
-        script=s,
+        script=script,
         inputs=kwtypes(f=CSVFile, y=FlyteDirectory, j=datetime.datetime),
-        output_locs=[OutputLocation(var="y", var_type=FlyteFile, location="{{ .inputs.f }}.pyc")],
+        output_locs=[OutputLocation(var="z", var_type=FlyteFile, location="{f}.pyc")],
     )
 
-    assert t.script == s
+    assert t.script == script
     y = t(f=test_csv, y=testdata, j=datetime.datetime(2021, 11, 10, 12, 15, 0))
     assert y.path[-4:] == ".pyc"
 
 
-def test_input_output_extra_var_in_template():
+@pytest.mark.parametrize(
+    "script",
+    [
+        (
+            """
+            cat {missing} >> {z}
+            echo "Hello World {y} on  {j} - output {x}"
+            """
+        ),
+        (
+            """
+            cat {f} {missing} >> {z}
+            echo "Hello World {y} on  {j} - output {x}"
+            """
+        ),
+    ],
+)
+def test_input_output_extra_and_missing_variables(script):
     t = ShellTask(
         name="test",
         debug=True,
-        script="""
-        cat {{ .inputs.f }} {{ .inputs.missing }} >> {{ .outputs.y }}
-        echo "Hello World {{ .inputs.y }} on  {{ .inputs.j }} - output {{.outputs.x}}"
-        """,
+        script=script,
         inputs=kwtypes(f=CSVFile, y=FlyteDirectory, j=datetime.datetime),
         output_locs=[
-            OutputLocation(var="x", var_type=FlyteDirectory, location="{{ .inputs.y }}"),
-            OutputLocation(var="y", var_type=FlyteFile, location="{{ .inputs.f }}.pyc"),
+            OutputLocation(var="x", var_type=FlyteDirectory, location="{y}"),
+            OutputLocation(var="z", var_type=FlyteFile, location="{f}.pyc"),
         ],
     )
 
-    with pytest.raises(ValueError):
+    with pytest.raises(ValueError, match="missing"):
         t(f=test_csv, y=testdata, j=datetime.datetime(2021, 11, 10, 12, 15, 0))
 
 
-def test_input_output_extra_input():
+def test_cannot_reuse_variables_for_both_inputs_and_outputs():
     t = ShellTask(
         name="test",
         debug=True,
         script="""
-        cat {{ .inputs.missing }} >> {{ .outputs.y }}
-        echo "Hello World {{ .inputs.y }} on  {{ .inputs.j }} - output {{.outputs.x}}"
+        cat {f} >> {y}
+        echo "Hello World {y} on  {j}"
         """,
         inputs=kwtypes(f=CSVFile, y=FlyteDirectory, j=datetime.datetime),
         output_locs=[
-            OutputLocation(var="x", var_type=FlyteDirectory, location="{{ .inputs.y }}"),
-            OutputLocation(var="y", var_type=FlyteFile, location="{{ .inputs.f }}.pyc"),
+            OutputLocation(var="y", var_type=FlyteFile, location="{f}.pyc"),
         ],
     )
 
-    with pytest.raises(ValueError):
+    with pytest.raises(ValueError, match="Variables {'y'} in Query"):
         t(f=test_csv, y=testdata, j=datetime.datetime(2021, 11, 10, 12, 15, 0))
 
 
+def test_can_use_complex_types_for_inputs_to_f_string_template():
+    @dataclass_json
+    @dataclass
+    class InputArgs:
+        in_file: CSVFile
```

Comment:
interesting, since this is using Flytefile, it will be automatically downloaded too?